src/reflow/mod.rs

```diff
@@ -0,0 +1,678 @@
+
+    let mut reflow_applied = false;
+    let mut lines = s_absolute.lines();
+    let mut indents = indentations.iter();
+    let last_ident =
+        indentations.last().expect("Must contain one indentation") - variant.prefix_len();
+    // first line has to be without indent and variant prefix
+    let (_, content, _) = gluon.next().expect("Must contain one line");
+    if lines.next() != Some(&content) {
+        reflow_applied = true;
+    }
+    let acc = content + &variant.suffix_string() + "\n";
+
+    // construct replacement string from prefix and Gluon iterations
+    let content = gluon.fold(acc, |mut acc, (_, content, _)| {
+        if lines.next() != Some(&content) {
+            reflow_applied = true;
+        }
+        let pre = if let Some(i) = indents.next() {
+            " ".repeat(*i - variant.prefix_len())
+        } else {
+            " ".repeat(last_ident)
+        };
+
+        acc.push_str(&pre);
+        acc.push_str(&variant.prefix_string());
+        acc.push_str(&content);
+        acc.push_str(&variant.suffix_string());
+        acc.push_str("\n");
+        acc
+    });
+
+    // remove last new line
+    let content = if let Some(c) = content.strip_suffix("\n") {
+        c.to_string()
+    } else {
+        return None;
+    };
+    // for MacroDocEq comments, we also have to remove the last closing delimiter
+    let content = if let Some(c) = content.strip_suffix(&variant.suffix_string()) {
+        c.to_string()
+    } else {
+        content
+    };
+
+    if reflow_applied {
+        Some(content)
+    } else {
+        None
+    }
+}
+
+/// Collect reflowed Paragraphs in a Vector of suggestions
+fn store_suggestion<'s>(
+    acc: &mut Vec<Suggestion<'s>>,
+    chunk: &'s CheckableChunk,
+    origin: &ContentOrigin,
+    paragraph: usize,
+    end: usize,
+    unbreakable_ranges: &[Range],
+    max_line_width: usize,
+) -> Result<usize> {
+    let range = Range {
+        start: paragraph,
+        end,
+    };
+    let mut spans = chunk.find_covered_spans(range.clone());
+    let span_start = if let Some(first) = spans.next() {
+        first
+    } else {
+        return Ok(paragraph);
+    };
+    let span_end = if let Some(last) = spans.last() {
+        last
+    } else {
+        span_start
+    };
+    let mut span = Span {
+        start: span_start.start,
+        end: span_end.end,
+    };
+    // TODO: find_covered_spans() seems to report a span which is off by one for TrppleSlash comments. Ultimately,
+    // the problem is somewhere inside chunk's source_mapping?!
+    if chunk.variant() == CommentVariant::TripleSlash
+        || chunk.variant() == CommentVariant::DoubleSlashEM
+    {
+        span.start.column += 1;
+    }
+
+    // Get indentation for each span, if a span covers multiple
+    // lines, use same indentation for all lines
+    let indentations = chunk
+        .find_covered_spans(range.clone())
+        .flat_map(|s| vec![s.start.column; s.end.line - s.start.line + 1])
```

Comment:
I'd be in favour of using the explicit `fn` for bound calculations, or additional `debug_assert` calls to make sure unit test catch the issues early on. `Spans` with reversed `start` and `end` are technically not an error state unless we define them as such by convention and assert as often as we can in unit tests.

Reply:
I've added a `debug_assert` for now. What exactly do you mean by explicit function? Something like `span.lines().collect()`?

Reply:
`saturating_sub` and friends